cpp/include/raft/sparse/matrix/preprocessing.cuh

```diff
@@ -0,0 +1,568 @@
+ * @param[in] coo_in
+ *   Raft container housing a compressed sparse row matrix representation.
+ * @param[in] results
+ *   array that is the size of the coo_in nnz that will store the encoded
+ *      transform values.
+ * @param[in] bm25_on
+ *   When true calculates bm25 encoding values, otherwise tfidf.
+ * @param[in] k_param
+ *   bm25 okapi optimization parameter k1
+ * @param[in] b_param
+ *   bm25 okapi optimization parameter b
+ * */
+template <typename ValueType, typename IndexType>
+void SparseEncoder<ValueType, IndexType>::transform(
+  raft::resources const& handle,
+  raft::device_coo_matrix<ValueType,
+                          IndexType,
+                          IndexType,
+                          IndexType,
+                          raft::device_uvector_policy,
+                          raft::PRESERVING> coo_in,
+  float* results,
+  bool bm25_on,
+  float k_param,
+  float b_param)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+  auto rows           = coo_in.structure_view().get_rows();
+  auto cols           = coo_in.structure_view().get_cols();
+  auto vals           = coo_in.view().get_elements();
+  auto nnz            = coo_in.structure_view().get_nnz();
+
+  auto d_rows = raft::make_device_vector<IndexType, int64_t>(handle, nnz);
+  auto d_cols = raft::make_device_vector<IndexType, int64_t>(handle, nnz);
+  auto d_vals = raft::make_device_vector<ValueType, int64_t>(handle, nnz);
+  raft::copy(d_rows.data_handle(), rows.data(), nnz, stream);
+  raft::copy(d_cols.data_handle(), cols.data(), nnz, stream);
+  raft::copy(d_vals.data_handle(), vals.data(), nnz, stream);
+
+  transform(
+    handle, d_rows.view(), d_cols.view(), d_vals.view(), nnz, results, bm25_on, k_param, b_param);
+}
+
+/**
+ * This function transforms the csr matrix based on statistics collected during fit
+ * cycle.
+ *
+ * @tparam ValueType
+ *   Type of the values in the sparse matrix.
+ * @tparam IndexType
+ *   Type of the indices associated with the values.
+ *
+ * @param[in] handle
+ *   Container for managing reusable resources.
+ * @param[in] csr_in
+ *   Raft container housing a compressed sparse row matrix representation.
+ * @param[in] results
+ *   array that is the size of the csr_in nnz that will store the encoded
+ *      transform values.
+ * @param[in] bm25_on
+ *   When true calculates bm25 encoding values, otherwise tfidf.
+ * @param[in] k_param
+ *   bm25 okapi optimization parameter k1
+ * @param[in] b_param
+ *   bm25 okapi optimization parameter b
+ * */
+template <typename ValueType, typename IndexType>
+void SparseEncoder<ValueType, IndexType>::transform(
+  raft::resources const& handle,
+  raft::device_csr_matrix<ValueType,
+                          IndexType,
+                          IndexType,
+                          IndexType,
+                          raft::device_uvector_policy,
+                          raft::PRESERVING> csr_in,
+  float* results,
+  bool bm25_on,
```

Comment:
if we have more options on algorithms, we may need an enum to avoid public API changing(after release)